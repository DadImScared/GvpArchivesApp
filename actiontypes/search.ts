export const UPDATE_SEARCH_QUERY = "search/UPDATE_SEARCH_QUERY";
export const TOGGLE_CATEGORIES = "search/TOGGLE_CATEGORIES";
export const SELECT_ALL_CATEGORIES = "search/SELECT_ALL_CATEGORIES";
export const REMOVE_ALL_CATEGORIES = "search/REMOVE_ALL_CATEGORIES";
