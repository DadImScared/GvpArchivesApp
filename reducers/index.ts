import { combineReducers } from "redux";

import audioPlayer, { IAudioPlayerState } from "./audioPlayer";
import itemsByCategory, { ItemsByCategoryState } from "./itemsByCategory";
import itemsById, { ItemsByIdState } from "./itemsById";
import loading, { ILoadingState } from "./loading";

export interface IReducerState {
  audioPlayer: IAudioPlayerState;
  loading: ILoadingState;
  itemsByCategory: ItemsByCategoryState;
  itemsById: ItemsByIdState;
}

const reducer = combineReducers<IReducerState>({
  audioPlayer,
  itemsByCategory,
  itemsById,
  loading,
});

export default reducer;
