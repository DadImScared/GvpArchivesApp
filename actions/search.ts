import axios, { AxiosPromise } from "axios";
import { AnyAction, Dispatch } from "redux";
import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/search";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { getQueryId } from "../reducers/search";
import autoComplete from "./autoComplete";
import { BASE_URL } from "./itemsById";

const autoCompleteUrl = `${BASE_URL}/api/v1/completeme/`;

export const isValidValidTimeToLive = (time: number) => time ? time >= +new Date() : false;

const updateSearchQuery = (query: string) => action(ActionTypes.UPDATE_SEARCH_QUERY, query);

type ApiCallCallback<ReturnType> = (
  state: IReducerState,
  dispatch: Dispatch
) => ReturnType;

type IAfterApiCall<Data> = (
  data: Data,
  state: IReducerState,
  dispatch: Dispatch
) => Promise<AnyAction[]> | AnyAction[];

export interface IApiCallConfig<Data> {
  afterApiCall?: IAfterApiCall<Data>;
  apiCall: ApiCallCallback<AxiosPromise<any>>;
  beforeApiCall?: ApiCallCallback<Promise<AnyAction[]> | AnyAction[]>;
  loadingId?: ((state: IReducerState) => string) | string;
  shouldCallApi?: ApiCallCallback<boolean | AxiosPromise<void>>;
}

export const getQuerySuggestions = (query: string, categories: string[]): IApiCallConfig<string[]> => {
  const queryId = getQueryId(query, categories);
  const loadingId = `${AUTOCOMPLETE_QUERY}${queryId}`;
  return {
    loadingId,
    afterApiCall: (data) => {
      return [
        autoComplete.addAutoComplete(
          query,
          categories,
          data.map((textObj: any) => textObj.text)
        )
      ];
    },
    apiCall: () => axios.get(`${autoCompleteUrl}${query}`),
    shouldCallApi: (state) => {
      const isLoading = state.loading[loadingId];
      if (!query || isLoading) return false;

      const currentAutoComplete = state.autoComplete[queryId];
      if (!currentAutoComplete) return true;

      return isValidValidTimeToLive(currentAutoComplete.timeToLive) ?
        false
        :
        axios.get(`${autoCompleteUrl}${query}`);
    }
  };
};

export default {
  updateSearchQuery
};
