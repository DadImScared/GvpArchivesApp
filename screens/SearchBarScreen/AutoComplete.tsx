import * as React from "react";
import { FlatList, ListRenderItem, Text, View } from "react-native";
import { NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";
import { compose } from "recompose";

import { Spinner } from "native-base";

import { IReducerState } from "../../reducers";
import { getAutoCompleteScreen, IAutoCompleteScreenData } from "../../reducers/autoComplete";

import ListItem from "../../components/SearchBarScreen/AutoCompleteListItem";

class AutoComplete extends React.Component<IAutoCompleteScreenData & NavigationScreenProps> {
  renderListItem: ListRenderItem<string> = ({ item }) => (
    <ListItem text={item} />
  );

  keyExtractor = (item: string, index: number) => `${item}-${this.props.queryId}-${index}`;

  render() {
    const { isLoading, query, autoComplete: { results } } = this.props;
    return (
      <View>
        {
          query ?
            isLoading || isLoading === undefined ?
              <Spinner color={"green"}/>
              :
              results.length ?
                <FlatList
                  keyExtractor={this.keyExtractor}
                  data={results}
                  renderItem={this.renderListItem}
                />
                :
                <Text>No suggestions</Text>
            :
            <Text>Type in the search bar to see suggestions!</Text>
        }
      </View>
    );
  }
}

const mapStateToProps = (state: IReducerState) => getAutoCompleteScreen(state);

const enhance = compose<any, any>(
  connect(mapStateToProps)
);

export default enhance(AutoComplete);
