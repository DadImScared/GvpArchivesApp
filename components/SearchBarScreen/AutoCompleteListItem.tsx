import * as React from "react";
import { Text } from "react-native";
import { NavigationScreenProp } from "react-navigation";

import { ListItem } from "native-base";

import { ISearchState } from "../../reducers/search";
import { searchParams } from "../../screens/Search";

interface IProps extends ISearchState {
  text: string;
  navigation: NavigationScreenProp<any>;
}

export class AutoCompleteListItem extends React.Component<IProps> {

  navigateSearch = () => {
    const { navigation, text, categories } = this.props;
    const parentNavigation = (navigation as any).dangerouslyGetParent().dangerouslyGetParent().dangerouslyGetParent();
    parentNavigation.replace("Search", searchParams(text, categories));
  };

  render() {
    return (
      <ListItem button={true} noIndent={true} onPress={this.navigateSearch}>
        <Text>{this.props.text}</Text>
      </ListItem>
    );
  }
}

export default AutoCompleteListItem;
