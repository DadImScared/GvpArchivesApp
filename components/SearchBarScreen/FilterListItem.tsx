import * as React from "react";
import { Platform, StyleSheet, View } from "react-native";

import { CheckBox, ListItem, NativeBase, Text } from "native-base";

import { SearchStateAndActions } from "../../reducers/search";
import { routes } from "../SideBar";

const styles = StyleSheet.create({
  innerListWrapper: {
    alignContent: "flex-start",
    alignItems: "flex-start",
    alignSelf: "flex-start",
    flexDirection: "row",
    justifyContent: "center",
    paddingRight: 0
  },
  text: {
    alignSelf: "center",
    marginLeft: 8,
    marginRight: 0,
    minWidth: 100
  }
});

interface IProps extends NativeBase.ListItem, SearchStateAndActions<"toggleCategory"> {
  category: string;
  addRow: (category: string, width: number) => void;
}

export class FilterListItem extends React.Component<IProps> {

  toggleCategory = Platform.select<any>({
    default: () => this.props.toggleCategory(this.props.category),
    ios: () => setTimeout(() => this.props.toggleCategory(this.props.category), 50)
  });

  addToRow = ({ nativeEvent: { layout: { width } } }: any) => {
    this.props.addRow(this.props.category, width + 36);
  };

  render() {
    const { category, toggleCategory, addRow, ...other } = this.props;
    return (
      <ListItem button={true} onPress={this.toggleCategory} {...other}>
        <View style={styles.innerListWrapper} onLayout={this.addToRow}>
          <CheckBox checked={other.selected} color={routes[category].icon.style.color} />
          <Text style={styles.text}>{routes[category].displayName}</Text>
        </View>
      </ListItem>
    );
  }
}

export default FilterListItem;
