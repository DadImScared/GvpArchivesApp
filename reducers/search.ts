import _ from "lodash";
import { ActionType } from "typesafe-actions";

import { search } from "../actions";
import * as ActionTypes from "../actiontypes/search";
import { IReducerState } from "./index";

export interface ISearchState {
  categories: string[];
  query: string;
}

export type SearchStateAndActions<
  T extends keyof ISearchState | keyof typeof search
  > = Pick<ISearchState & typeof search, T>;

export const CATEGORIES = [
  "bhagavatpatrika",
  "book",
  "harikatha",
  "harmonistmagazine",
  "harmonistmonthly",
  "lecture",
  "movie",
  "song"
];

export const getInitialSearchState = (): ISearchState => ({
  categories: CATEGORIES,
  query: ""
});

export default function(state = getInitialSearchState(), action: ActionType<typeof search>) {
  switch (action.type) {
    case ActionTypes.UPDATE_SEARCH_QUERY:
      return {
        ...state,
        query: action.payload
      };
    case ActionTypes.TOGGLE_CATEGORIES:
      const { payload } = action;
      const { categories } = state;
      return {
        ...state,
        categories: categories.includes(payload) ?
          categories.filter((category) => category !== action.payload)
          :
          [...categories, payload]
      };
    case ActionTypes.SELECT_ALL_CATEGORIES:
      return {
        ...state,
        categories: [...CATEGORIES]
      };
    case ActionTypes.REMOVE_ALL_CATEGORIES:
      return {
        ...state,
        categories: []
      };
    default:
      return state;
  }
}

export const getQueryId = (query: string, categories: string[]) => {
  return `${query}-${_.sortBy(categories, (o) => o).join(",")}`;
};

export const getSearch = (state: IReducerState) => state.search;
