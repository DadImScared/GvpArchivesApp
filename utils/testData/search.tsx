import * as React from "react";
import * as renderer from "react-test-renderer";

import { action } from "@storybook/addon-actions";
import _ from "lodash";

import { getQueryId } from "../../reducers/search";

export const makeTree = (Component: any, props: any = {}) => (injectedProps: any = {}) => {
  return renderer.create(<Component {...props} {...injectedProps} />);
};

const params = {
  categories: ["book", "song"],
  query: "radha",
  queryId: getQueryId("radha", ["book", "song"])
} as any;

export const getMockFunction = ({
  storybook = false,
  actionName = "",
  cb = false
}: any) => {
  if (storybook) {
    return (...args: any) => {
      action(actionName)(...args);
      if (cb) {
        return cb(...args);
      }
    };
  }
  return cb ? jest.fn(cb) : jest.fn();
};

export const searchTestProps = ({ props = {}, storybook = false } = {}) => (_.merge({
  addRecentSearch: getMockFunction({ storybook, actionName: "addRecentSearch" }),
  getSearchResults: getMockFunction({ storybook, actionName: "getSearchResults" }),
  isLoading: false,
  navigation: {
    getParam: getMockFunction({ storybook, actionName: "getParam", cb: (param: string) => params[param] }),
    push: getMockFunction({ storybook, actionName: "push" })
  },
  results: [
    {
      category: "book",
      item_id: "item-1",
      link: "item-1-link",
      title: "item-1-title"
    }
  ],
  suggestions: ["Suggestion-1", "Suggestion-2", "Suggestion-3"],
  timeToLive: 12345
}, props));
