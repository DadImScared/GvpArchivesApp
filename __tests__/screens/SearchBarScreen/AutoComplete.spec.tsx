import * as React from "react";
import * as renderer from "react-test-renderer";

import { ListItem } from "native-base";

import { getQueryId } from "../../../reducers/search";
import { AutoComplete } from "../../../screens/SearchBarScreen/AutoComplete";

describe("AutoComplete", () => {
  it("should navigate onPress list item", () => {
    const props = {
      autoComplete: {
        results: ["suggestion-1"],
        timeToLive: 1234234
      },
      categories: ["book"],
      isLoading: false,
      navigation: {
        dangerouslyGetParent: () => props.navigation,
        replace: jest.fn()
      } as any,
      query: "radha",
      queryId: getQueryId("radha", ["book"]),
      updateSearchQuery: jest.fn()
    };
    const tree = renderer.create(<AutoComplete {...props} />);
    tree.root.findByType(ListItem).props.onPress();
    expect(props.navigation.replace.mock.calls).toMatchSnapshot("navigation args");
    expect(props.updateSearchQuery.mock.calls).toMatchSnapshot("updateSearchQuery args");
  });
});
