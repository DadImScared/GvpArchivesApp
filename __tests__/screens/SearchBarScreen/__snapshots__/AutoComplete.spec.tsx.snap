// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete should navigate onPress list item: navigation args 1`] = `
Array [
  Array [
    "Search",
    Object {
      "categories": Array [
        "book",
      ],
      "query": "suggestion-1",
      "queryId": "suggestion-1-book",
    },
  ],
]
`;

exports[`AutoComplete should navigate onPress list item: updateSearchQuery args 1`] = `
Array [
  Array [
    "suggestion-1",
  ],
]
`;
